src/neo/SmartContract/Native/Tokens/Nep11Token.cs

```diff
@@ -0,0 +1,269 @@
+            storages.Add(CreateOwnershipKey(Prefix_OwnerToTokenId, account, innerKey), new StorageItem(new TAccount() { Balance = Factor }));
+            storages.Add(CreateOwnershipKey(Prefix_TokenIdToOwner, innerKey, account), new StorageItem(new byte[0]));
+
+            IncreaseTotalSupply(engine.Snapshot);
+
+            engine.SendNotification(Hash, new Array(new StackItem[] { "Transfer", StackItem.Null, account.ToArray(), Factor, token.Id }));
+        }
+
+        internal protected virtual void Burn(ApplicationEngine engine, byte[] tokenId)
+        {
+            var storages = engine.Snapshot.Storages;
+            UInt160 innerKey = GetInnerKey(tokenId);
+            StorageKey tokenKey = CreateTokenKey(innerKey);
+            if (!storages.Delete(tokenKey)) throw new InvalidOperationException("Token doesn't exist");
+
+            IEnumerator enumerator = OwnerOf(engine.Snapshot, tokenId);
+            if (enumerator.Next())
+            {
+                UInt160 owner = new UInt160(enumerator.Value().GetSpan().ToArray());
+                storages.Delete(CreateOwnershipKey(Prefix_TokenIdToOwner, innerKey, owner));
+                storages.Delete(CreateOwnershipKey(Prefix_OwnerToTokenId, owner, innerKey));
+            }
+
+            DecreaseTotalSupply(engine.Snapshot);
+
+            engine.SendNotification(Hash, new Array(new StackItem[] { "Transfer", StackItem.Null, StackItem.Null, Factor, tokenId }));
+        }
+
+        public virtual UInt160 GetInnerKey(byte[] tokenId)
+        {
+            return new UInt160(Crypto.Hash160(tokenId));
+        }
+
+        private StorageKey CreateOwnershipKey(byte prefix, UInt160 first, UInt160 second)
+        {
+            byte[] byteSource = new byte[first.Size + second.Size];
+            System.Array.Copy(first.ToArray(), 0, byteSource, 0, first.Size);
+            System.Array.Copy(second.ToArray(), 0, byteSource, first.Size, second.Size);
+            return CreateStorageKey(prefix, byteSource);
+        }
+
+        protected StorageKey CreateTokenKey(UInt160 innerKey)
+        {
+            return CreateStorageKey(Prefix_TokenId, innerKey.ToArray());
+        }
+
+        private void IncreaseTotalSupply(StoreView snapshot)
+        {
+            StorageItem storage_totalSupply = snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_TotalSupply), () => new StorageItem() { Value = BigInteger.Zero.ToByteArray() });
+            storage_totalSupply.Value = (new BigInteger(storage_totalSupply.Value) + BigInteger.One).ToByteArray();
+        }
+
+        private void DecreaseTotalSupply(StoreView snapshot)
+        {
+            StorageItem storage_totalSupply = snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_TotalSupply), () => new StorageItem() { Value = BigInteger.Zero.ToByteArray() });
+            BigInteger totalSupply = new BigInteger(storage_totalSupply.Value);
+            if (totalSupply.Equals(BigInteger.Zero)) return;
```

Comment:
Exception?

Reply:
The total cannot be negative.So when the result is 0, it will not continue to decrease

Reply:
If throw exception, it'll stop other assets to be burned.